topi/python/topi/cuda/conv2d_winograd.py

```diff
@@ -0,0 +1,372 @@
+
+    if tile_size == 4:
+        G_data = np.array([
+            [1 / 4.0, 0, 0],
+            [-1 / 6.0, -1 / 6.0, -1 / 6.0],
+            [-1 / 6.0, 1 / 6.0, -1 / 6.0],
+            [1 / 24.0, 1 / 12.0, 1 / 6.0],
+            [1 / 24.0, -1 / 12.0, 1 / 6.0],
+            [0, 0, 1]], dtype=np.float32)
+
+        B_data = np.array([
+            [4, 0, 0, 0, 0, 0],
+            [0, -4, 4, -2, 2, 4],
+            [-5, -4, -4, -1, -1, 0],
+            [0, 1, -1, 2, -2, -5],
+            [1, 1, 1, 1, 1, 0],
+            [0, 0, 0, 0, 0, 1]], out_dtype)
+
+        A_data = np.array([
+            [1, 0, 0, 0],
+            [1, 1, 1, 1],
+            [1, -1, 1, -1],
+            [1, 2, 4, 8],
+            [1, -2, 4, -8],
+            [0, 0, 0, 1]], out_dtype)
+    elif tile_size == 2:
+        G_data = np.array([
+            [1, 0, 0],
+            [1.0/2, 1.0/2, 1.0/2],
+            [1.0/2, -1.0/2, 1.0/2],
+            [0, 0, 1]], np.float32)
+
+        B_data = np.array([
+            [1, 0, 0, 0],
+            [0, 1, -1, 1],
+            [-1, 1, 1, 0],
+            [0, 0, 0, -1]], out_dtype)
+
+        A_data = np.array([
+            [1, 0],
+            [1, 1],
+            [1, -1],
+            [0, -1]], out_dtype)
+    else:
+        raise ValueError("Unsupported tile size for winograd: " + str(tile_size))
+
+    m = A_data.shape[1]
+    r = 3
+    alpha = m + r - 1
+    H = (H + 2 * HPAD - KH) // HSTR + 1
+    W = (W + 2 * WPAD - KW) // WSTR + 1
+    nH, nW = (H + m-1) // m, (W + m-1) // m
+    P = N * nH * nW
+
+    # transform kernel
+    if not pre_computed:
+        G = const_matrix(G_data, 'G')
+        r_kh = tvm.reduce_axis((0, KH), name='r_kh')
+        r_kw = tvm.reduce_axis((0, KW), name='r_kw')
+        kernel_pack = tvm.compute((alpha, alpha, CO, CI), lambda eps, nu, co, ci:
+                                  tvm.sum(kernel[co][ci][r_kh][r_kw] *
+                                          G[eps][r_kh] * G[nu][r_kw],
+                                          axis=[r_kh, r_kw]), name='kernel_pack')
+    else:
+        kernel_pack = kernel
+
+    # pack input tile
+    input_tile = tvm.compute((CI, P, alpha, alpha), lambda c, p, eps, nu:
+                             data_pad[p // (nH * nW)][c][p // nW % nH * m + eps]
+                             [p % nW * m + nu], name='d')
+
+    # transform data
+    B = const_matrix(B_data)
+    r_a = tvm.reduce_axis((0, alpha), 'r_a')
+    r_b = tvm.reduce_axis((0, alpha), 'r_a')
+    data_pack = tvm.compute((alpha, alpha, CI, P), lambda eps, nu, ci, p:
+                            tvm.sum(input_tile[ci][p][r_a][r_b] * B[r_a][eps] * B[r_b][nu],
+                                    axis=[r_a, r_b]), name='data_pack')
+
+    # do batch gemm
+    ci = tvm.reduce_axis((0, CI), name='ci')
+    bgemm = tvm.compute((alpha, alpha, CO, P), lambda eps, nu, co, p:
+                        tvm.sum(kernel_pack[eps][nu][co][ci] *
+                                data_pack[eps][nu][ci][p],
```

Comment:
Are you sure there is no bank conflict when accessing data_pack here?

Reply:
I thinks the bank conflict happens for kernel_pack. I should change the layout of kernel_pack to [eps][nu][ci][co]

Reply:
That seems right, I used the same memory layout [here](https://github.com/masahi/tvm-winograd/blob/master/wino_test_cuda.py#L185-L187) and so does the [gemm recipe](https://github.com/dmlc/tvm/blob/master/topi/recipe/gemm/cuda_gemm_square.py#L41).